### PR TITLE
PyTorch 使用层（上）：二十行训练循环改用代码引用（!ref）——#49 的另一种答法

### PR DESCRIPTION
**Pull request by arganzheng**
## Summary
- 《算法工程师的工具箱（03）：PyTorch 使用层（上）》第六章的二十行训练循环：去掉代码里的 ①–⑯ 标号，改成 `# !ref` 锚点（13 个：model / opt / sched / batch / to-dev / autocast / forward / loss / backward / clip / update / log / eval）；「逐行解释」每条开头、「多出来的四样」「一个 CPU 上的陷阱」「与 Trainer 的关系」里的引用改成链接——悬停亮行、点击跳转，代码左侧蓝色行号悬停弹出解释。
- 系列总结（`algorithm-tooling-series-recap-and-self-test`）过时的「`compute_loss` 是第 9–10 行、`training_step` 是第 11–13 行」改链到那两组行（跨文锚点，lychee 校验）。

来源：issue #49（@magiforrest 划线「compute_loss 是第 9–10 行…」：「搞啥飞机，完全不知道这里是在哪里」——当时用标号修过一次，这次用新的代码引用功能重做）。

#### Test plan
- [x] `jekyll build`：`pre[data-refs]` 含全部 13 个名字，`data-lines=24`，26 个正文引用全部有对应行
- [x] 浏览器：悬停 / 点击 / 行号槽气泡
- [x] pre-push `npm run check` 全绿（含跨文 `#forward` / `#backward` 锚点）

Generated with [Devin](https://devin.ai)

### CODE REVIEW

**Review by arganzheng — COMMENTED** [review]
修改理由（每节一条，由 tools/review.py --notes 发出）

File: _posts/2026-01-19-pytorch-in-use-five-objects-and-a-training-loop.md [hunk 1]
```diff
@@ -382,57 +382,70 @@ def get_batch(data, cfg, gen):
 cfg = Config()
 torch.manual_seed(cfg.seed); gen = torch.Generator().manual_seed(cfg.seed)
 train, val = load_corpus()
-model = TinyGPT(cfg).to(DEV)                                                       # ①
-opt = torch.optim.AdamW(model.parameters(), lr=cfg.lr, weight_decay=0.1)           # ②
-sched = torch.optim.lr_scheduler.LambdaLR(opt, lambda s: cosine_with_warmup(s, cfg))  # ③
+# !ref model
```
Comment (arganzheng) [inline _posts/2026-01-19-pytorch-in-use-five-objects-and-a-training-loop.md:385]: 代码里的 `# ①`…`# ⑯` 行尾标号全部去掉，换成 13 个独占一行的 `# !ref` 锚点（渲染时被删，行数从 25 变 24 是因为 `# [B, T, V]` 那行的对齐空格也收了）。⑪⑫⑬ 三个动作在一行，合成一个 `update`；`log +1` 和 `eval +4` 各覆盖多行。块下那句「第 ④ 行 / 第 ⑧ 行」改成链接。

File: _posts/2026-01-19-pytorch-in-use-five-objects-and-a-training-loop.md [hunk 2]
```diff
@@ -382,57 +382,70 @@ def get_batch(data, cfg, gen):
...
-- **⑨** `loss.backward()`：反向传播，梯度累加到每个参数的 `.grad`（第三章第二件事；L0 第七篇链式法则）。
-- **⑩** `clip_grad_norm_(..., 1.0)`：算所有梯度拼起来的总范数，超过 1.0 就整体缩放到 1.0——**梯度裁剪**，防 loss spike（L3 第三篇）。返回值是裁剪前的范数，值得打出来看。
-- **⑪** `opt.step()`：AdamW 用 `.grad` 更新参数（L3 第三篇）。
-- **⑫** `sched.step()`：学习率按曲线走一步（L0 第七篇第六章）。
-- **⑬** `opt.zero_grad(set_to_none=True)`：清梯度；置 `None` 而不是填 0，省一次显存写。放在 `step` 之后而不是 `backward` 之前，是为了让第三章说的梯度累积只改这一行的位置就能实现。
-- **⑭** 每 100 步才 `loss.item()`：它会同步 GPU（第二章第 3 节），别每步做。
-- **⑮ ⑯** 评估：`model.eval()` 切换 dropout / BatchNorm 的行为（本模型没有，写上是习惯），`torch.no_grad()` 不建图（第三章第三件事），验证集上算一次 loss，`exp` 一下就是 [PPL](# "tip: perplexity，困惑度 = exp(每 token 交叉熵)。直觉是模型每步在多少个候选里犹豫：PPL 128 等于在 128 个字符里均匀乱猜，PPL 9.6 是在不到 10 个里犹豫；L0 第六篇")。
+下面每条开头的代码就是上面的某一行：悬停它，那一行会亮起；点击跳过去。反过来，代码左侧蓝色的行号悬停就能看到对应的解释。
```
Comment (arganzheng) [inline _posts/2026-01-19-pytorch-in-use-five-objects-and-a-training-loop.md:425]: 每条开头从粗体标号改成那一行代码本身的链接——悬停亮行、点击跳过去；⑪⑫⑬ 合成一条（同一行的三件事），⑮⑯ 合成「评估那五行」。加了一句开头说明怎么用。

File: _posts/2026-01-19-pytorch-in-use-five-objects-and-a-training-loop.md [hunk 3]
```diff
@@ -382,57 +382,70 @@ def get_batch(data, cfg, gen):
...
 
 ### 4. 多出来的四样
 
 它比教程里常见的循环长。教程的最小版只有五行——前向、算 loss、`backward`、`step`、`zero_grad`——那是[第一章](#一总览)那个环的骨架，能跑 MNIST。多出来的四样是 LLM 训练的标配，每一样都对应一种不加就会遇到的故障：
 
-1. **`autocast`（第 ⑥ 行）。** 不开，所有矩阵乘在 fp32 上跑：显存里的激活是 bf16 的两倍，Tensor Core 的 bf16 吞吐也用不上，同一张卡上速度与能放的 batch 都差一倍多。只对 GPU 成立——CPU 上开它是纯开销。
-2. **`.float()` 与 `ignore_index`（第 ⑧ 行）。** 不加 `.float()`，softmax 在 bf16 上算：bf16 只有 8 位尾数，几万个 logits 里 `exp` 之后求和会丢掉小项，loss 从第一步起就带着系统误差，训到后期梯度不准。不加 `ignore_index=-100`，SFT 数据里的 prompt 也被当成学习目标——模型花一半算力去学"复述问题"，而且 padding 位置的 loss 会把平均值拉偏。
-3. **`clip_grad_norm_`（第 ⑩ 行）。** 不裁，某一个 batch 里的坏样本产生一个特别大的梯度，一步就把参数推到很远的地方，loss 冲上去，之后可能回不来——这就是 loss spike。裁剪把这一步的总范数压回 1.0，参数只往那个方向走一小步。
-4. **学习率调度（第 ③、⑫ 行）。** 不 warmup，第一步就用 $$3 \times 10^{-4}$$：AdamW 的二阶矩估计在前几步还没稳定，实际步长可能比设定大很多倍，前几步就发散（L3 第三篇讲 Adam 为什么需要 warmup）；不衰减，后期学习率太大，loss 在最优点附近来回抖、收不下去。
+1. **[`autocast`](#autocast)。** 不开，所有矩阵乘在 fp32 上跑：显存里的激活是 bf16 的两倍，Tensor Core 的 bf16 吞吐也用不上，同一张卡上速度与能放的 batch 都差一倍多。只对 GPU 成立——CPU 上开它是纯开销。
```
Comment (arganzheng) [inline _posts/2026-01-19-pytorch-in-use-five-objects-and-a-training-loop.md:445]: 四处「（第 ⑥ 行）」式引用改成链接；「学习率调度（第 ③、⑫ 行）」拆成「建调度器」「每步 `sched.step()`」两个链接，因为它们是两行。

File: _posts/2026-01-19-pytorch-in-use-five-objects-and-a-training-loop.md [hunk 4]
```diff
@@ -462,11 +475,11 @@ step  999  loss 2.022  lr 3.00e-05  grad_norm 0.31   53.6s
 
 ### 6. 一个 CPU 上的陷阱
 
-第 ⑥ 行的 `enabled=DEV == "cuda"` 不是可有可无的：在 CPU 上开着 bf16 autocast 训，一步 1.4 秒；关掉是 0.05 秒——**慢 30 倍**。原因是 CPU 没有 bf16 的硬件路径，PyTorch 用软件模拟。混合精度的收益完全来自硬件（GPU 的 Tensor Core），没有硬件时它只是开销。第四篇讲它在 GPU 上为什么快、省多少显存。
+[`autocast` 那一行](#autocast)的 `enabled=DEV == "cuda"` 不是可有可无的：在 CPU 上开着 bf16 autocast 训，一步 1.4 秒；关掉是 0.05 秒——**慢 30 倍**。原因是 CPU 没有 bf16 的硬件路径，PyTorch 用软件模拟。混合精度的收益完全来自硬件（GPU 的 Tensor Core），没有硬件时它只是开销。第四篇讲它在 GPU 上为什么快、省多少显存。
```
Comment (arganzheng) [inline _posts/2026-01-19-pytorch-in-use-five-objects-and-a-training-loop.md:478]: 「第 ⑥ 行的」改成「`autocast` 那一行的」链接。

File: _posts/2026-01-19-pytorch-in-use-five-objects-and-a-training-loop.md [hunk 5]
```diff
@@ -462,11 +475,11 @@ step  999  loss 2.022  lr 3.00e-05  grad_norm 0.31   53.6s
...
 
 ### 7. 与 `Trainer` 的关系
 
-`transformers.Trainer`、`trl.SFTTrainer` 做的是同样的事，对应到上面的标号：`compute_loss` 是第 ⑦–⑧ 行（前向 + 算 loss），`training_step` 是第 ⑨–⑬ 行（反向、裁剪、更新、清零），`autocast` 由 `TrainingArguments(bf16=True)` 打开，学习率曲线由 `lr_scheduler_type` 与 `warmup_steps` 决定，外面再包上日志、评估、checkpoint 与分布式。它们的行为不符合预期时——loss 不降、显存爆、学习率不对——回到这二十行想"它在第几行做了和我不一样的事"，然后去读它的源码（第五篇给入口）。
+`transformers.Trainer`、`trl.SFTTrainer` 做的是同样的事，对应到上面的代码：`compute_loss` 是[前向](#forward) + [算 loss](#loss)两行，`training_step` 是[反向](#backward)、[裁剪](#clip)、[更新与清零](#update)三行，`autocast` 由 `TrainingArguments(bf16=True)` 打开，学习率曲线由 `lr_scheduler_type` 与 `warmup_steps` 决定，外面再包上日志、评估、checkpoint 与分布式。它们的行为不符合预期时——loss 不降、显存爆、学习率不对——回到这二十行想"它在第几行做了和我不一样的事"，然后去读它的源码（第五篇给入口）。
```
Comment (arganzheng) [inline _posts/2026-01-19-pytorch-in-use-five-objects-and-a-training-loop.md:482]: #49 划的就是这句。「第 ⑦–⑧ 行 / 第 ⑨–⑬ 行」改成「前向 + 算 loss 两行」「反向、裁剪、更新与清零三行」，每个词链到对应行。

File: _posts/2026-01-22-algorithm-tooling-series-recap-and-self-test.md [hunk 6]
```diff
@@ -156,7 +156,7 @@ date: 2026-01-22 20:00:00
 
 ### 4. 从封装回到二十行：追溯的路径
 
-总纲把第三种能力叫追溯：从高层封装回到二十行，从报错回到四块，从慢回到 profiler 表。第一篇给第一层地图：PyTorch 的每个 API 建在哪个 Python 协议上，traceback 从下往上读、跳过 `_call_impl`、找自己文件的最后一帧。第三篇给第二层：二十行的每一行对应一个概念，`Trainer` 的 `compute_loss` 是第 9–10 行、`training_step` 是第 11–13 行，行为不对时回到这张表想"它在哪一行做了不同的事"。第五篇给第三层：六行 SFT 背后的六件事各落在二十行的哪一行（chat template 在 `__getitem__`、loss mask 在 `ignore_index=-100`、packing 在 `collate_fn`、LoRA 在 `requires_grad`），然后从 `compute_loss` 往下追源码——答案学会了却不会停这种问题只有读源码才能定位。
+总纲把第三种能力叫追溯：从高层封装回到二十行，从报错回到四块，从慢回到 profiler 表。第一篇给第一层地图：PyTorch 的每个 API 建在哪个 Python 协议上，traceback 从下往上读、跳过 `_call_impl`、找自己文件的最后一帧。第三篇给第二层：二十行的每一行对应一个概念，`Trainer` 的 `compute_loss` 是[前向 + 算 loss 那两行](/pytorch-in-use-five-objects-and-a-training-loop.html#forward)、`training_step` 是[反向到清零那三行](/pytorch-in-use-five-objects-and-a-training-loop.html#backward)，行为不对时回到这张表想"它在哪一行做了不同的事"。第五篇给第三层：六行 SFT 背后的六件事各落在二十行的哪一行（chat template 在 `__getitem__`、loss mask 在 `ignore_index=-100`、packing 在 `collate_fn`、LoRA 在 `requires_grad`），然后从 `compute_loss` 往下追源码——答案学会了却不会停这种问题只有读源码才能定位。
```
Comment (arganzheng) [inline _posts/2026-01-22-algorithm-tooling-series-recap-and-self-test.md:159]: 这句还停在第一版的「第 9–10 行、第 11–13 行」（第一次修 #49 时没有同步），现在改成跨文链接到那篇的 `#forward` / `#backward` 行锚点——链接检查会盯着它们。

**Review by arganzheng — COMMENTED** [review]
修改理由（每节一条，由 tools/review.py --notes 发出）

File: _posts/2026-09-09-blog-memo.md [hunk 7]
```diff
@@ -279,7 +279,12 @@ kramdown（GFM 模式），加上博客自己的扩展：
 - **解释性脚注不受影响**：名字不要以 `q` 开头（`[^nccl]`、`[^ring]` 这样即可），继续浮窗。所以一篇文章里两种脚注可以并存，只靠名字区分。
 - 写法：答案一到几句、可核对、含数字，末尾「详见 `[第 N 章](#锚点)`」链到章标题（锚点是 kramdown 生成的 `<h2 id>`，从 `_site` 里抄）；定义放「下一篇」之后即可，kramdown 渲染时本来就把脚注放在正文末尾。
 
-**代码引用**：把下面的一段说明和代码块里的某几行绑起来。代码里在要指的那行**上面**单独写一行注释 `# !ref 名字`（`+N` 再多包 N 行；`//`、`--`、`;`、`/* */`、`<!-- -->` 等注释符都认），正文用普通链接 `[文字](#名字)` 指过去。看效果——悬停下面的划线文字，或悬停代码左侧蓝色的行号：
+**代码引用**（逐行讲解代码时用）：把正文里的一段说明和代码块里的某几行**绑起来**，读者从任何一边都能找到另一边。以前的做法是在代码注释里写 ①②③、下面列表逐条对应——读者还是得自己在两边找数字，而且 ① 会跟着代码一起被复制走。现在两步：
```
Comment (arganzheng) [inline _posts/2026-09-09-blog-memo.md:282]: 「代码引用」一段按要求写清楚：先说为什么（①②③ 要读者自己两边找数字、还会被复制走），再是两步写法、页面上两个方向各发生什么、五条要记的规则，末尾链到第一篇用上它的文章（本 PR 里的 PyTorch 使用层（上））对着看。

**Review by arganzheng — COMMENTED** [review]
修改理由（每节一条，由 tools/review.py --notes 发出）
